web: add API functions for retrieving unique values of specific keys

sumpfralle opened the pull request:
the following API paths are added:
* /item/unique/FIELD
* /album/unique/FIELD

Both paths will deliver a json dictionary with the key 'values' containing a
list of unique values belonging to the requested table and field.

This feature is useful for categorization by remote clients, e.g. Mopidy-Beets:
* track genres: /item/unique/genre
* track languages: /album/unique/language
* album artists: /album/unique/albumartist
* album years: /album/unique/year


Notes for the reviewer: please take a look at my usage of the half-private attributes "_table" and "_fields". I thought I could use "all_keys()" for the latter, but this attribute was not accessible. I do not know of an alternative for "_table".
Is this attribute access acceptable for you?

sampsyo commented:
Cool! This looks reasonable to me. Thanks for thinking of it!

Using the private `_table` field is of course fine, since querying SQL directly is the right thing to do here. But I'm a little confused about why `all_keys` didn't work. We use `Item.all_keys()` and `Album.all_keys()` in a few places. What was the trouble you were running into?

And some obvious bikeshedding that doesn't really matter, so feel free to ignore it:

* The simpler endpoint name `/values/` might be slightly more descriptive than `/unique/`.
* Similarly, using `/unique/key?sort=sort_key` might make a little more sense here than `/unique/key/sort_key` since there's not really a hierarchy of resources here.

Anyway, these mostly don't matter because this is a beets-specific API. In the future, though, the [AURA][] project would like to make these things a little more standardized, at which point it could be a good idea to be careful about conventions.

[aura]: https://github.com/beetbox/aura

sumpfralle commented:
Thank you for sharing your bikeshedding inspirations. I liked and implemented all of them.
The "all_keys" method was missing for me, since my local installation is bit older. I fixed these references, as well.

sampsyo commented:
Awesome! This looks fantastic. I'll merge it now. Thanks for adding it, and for being patient with my suggestions! :rocket:
